Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs

```diff
@@ -0,0 +1,336 @@
+    {
+        base.Started(uid, component, gameRule, args);
+    }
+    private void AssignHeadRevs(RevolutionaryRuleComponent comp)
+    {
+        var allPlayers = _playerSystem.ServerSessions.ToList();
+        var playerList = new List<IPlayerSession>();
+        var prefList = new List<IPlayerSession>();
+        var cantBeAntag = new List<IPlayerSession>();
+        foreach (var player in allPlayers)
+        {
+            var mind = player.GetMind();
+            if (!player.Data.ContentData()?.Mind?.AllRoles.All(role => role is not Job { CanBeAntag: false }) ?? false)
+            {
+                cantBeAntag.Add(player);
+                continue;
+            }
+            if (player.AttachedEntity == null || HasComp<HumanoidAppearanceComponent>(player.AttachedEntity))
+            {
+                playerList.Add(player);
+            }
+            else
+                continue;
+
+            var pref = (HumanoidCharacterProfile) _prefs.GetPreferences(player.UserId).SelectedCharacter;
+            if (pref.AntagPreferences.Contains(comp.HeadRevPrototypeId))
+                prefList.Add(player);
+        }
+        if (playerList.Count == 0)
+            return;
+
+        _sawmill.Error("Math");
+        var headRevs = Math.Clamp(allPlayers.Count / HeadRevsPerPlayer, 1, MaxRevHeads);
+        for (var revs = 0; revs < headRevs; revs++)
+        {
+            IPlayerSession headRev;
+            if (prefList.Count == 0)
+            {
+                if (playerList.Count == 0)
+                {
+                    break;
+                }
+                headRev = _random.PickAndTake(playerList);
+            }
+            else
+            {
+                headRev = _random.PickAndTake(prefList);
+                playerList.Remove(headRev);
+            }
+            var mind = headRev.Data.ContentData()?.Mind;
+            if (mind == null)
+            {
+                continue;
+            }
+            _mindSystem.AddRole(mind, new RevolutionaryRole(mind, _prototypeManager.Index<AntagPrototype>(comp.HeadRevPrototypeId)));
+            var inCharacterName = string.Empty;
+            if (mind.OwnedEntity != null)
+            {
+                AddComp<RevolutionaryRuleComponent>(mind.OwnedEntity.Value);
+                AddComp<HeadRevolutionaryComponent>(mind.OwnedEntity.Value);
+                AddComp<RevolutionaryComponent>(mind.OwnedEntity.Value);
+                _stationSpawningSystem.EquipStartingGear(mind.OwnedEntity.Value, _prototypeManager.Index<StartingGearPrototype>(comp.HeadRevGearPrototypeId), null);
+                _npcFaction.RemoveFaction(mind.OwnedEntity.Value, "NanoTrasen", false);
+                _npcFaction.AddFaction(mind.OwnedEntity.Value, "Revolutionary");
+                inCharacterName = MetaData(mind.OwnedEntity.Value).EntityName;
+                comp.HeadRevs.Add(inCharacterName, headRev.Name);
```

Comment:
This should probably be pulled into it's own function on this system to avoid duplication with the flash's code.